feat: Show bluesheets using Attended tables

pselkirk opened the pull request:
Fixes #6898
Fixes #6454

pselkirk committed:
feat: Show bluesheets using Attended tables (#6898)

pselkirk committed:
feat: Allow users to add themselves to session attendance (#6454)

jennifer-richards reviewed (changes requested):
I think there are a couple of sharp corners that should be fixed - at least putting a check before accepting a self-declared attendance POST is a must-fix.

pselkirk committed:
chore: Correct copyright year

pselkirk commented on ietf/meeting/models.py:
```diff
@@ -1479,6 +1479,8 @@ class Meta:
     person = ForeignKey(Person)
     session = ForeignKey(Session)
+    time = models.DateTimeField(auto_now_add=True, null=True, blank=True)
+    origin = models.CharField(max_length=32, default='datatracker')
```
TBH, I copied these fields from `Volunteer`. But there appears to be no mechanism for setting `Volunteer.origin` to anything other than the default, and all 1076 Volunteer objects have origin='datatracker'. So I don't know what that's all about.

In any case, `Attendance.origin` is not intended to be a user-visible field - 'self declared' is an indication that the record was created as an action of the "I was there" button.

jennifer-richards replied:
It still worries me a bit to have a free-form `CharField` that is expected to have only specific values. Given that even using `choices` provides minimal protection, I'd be content with a comment here just so people don't have to search the code to find out what values to consider.

rjsparks replied:
Volunteer can also be set by the registration system (and I thought that was already happening - I will followup).

rjsparks replied:
It _will_ be set with registration for 120

pselkirk commented on ietf/meeting/utils.py:
```diff
@@ -155,6 +156,8 @@ def finalize(meeting):
             else:
                 sp.rev = '00'
             sp.save()
+        if meeting.number >= '118':
```
Duh, that was a last-minute addition, and it was late. Up to a minute before I committed, `generate_bluesheet` was unconditional. The original ticket said to start giving the live view of attendance with 118, but we later changed that to any un-finalized meeting. The intent of this check was to avoid overwriting good pre-existing bluesheet files in the unlikely event that, say, we had to un-finalize and re-finalize 107 (before we had `Attended` records). But that's pretty far-fetched, and maybe I should just remove that check.

jennifer-richards replied:
I agree it's unlikely, but I like the idea of a safeguard.

pselkirk replied:
Nah, keep the check, but roll it back to 108 (the earliest `Attended` records).

pselkirk commented on ietf/meeting/views.py:
```diff
@@ -2547,10 +2547,16 @@ def session_attendance(request, session_id, num):
 
+    person = request.user.person
+    if request.method=='POST':
+        session.attended_set.get_or_create(person=person, origin="self declared")
```
First point is totally legit, but I don't understand the second point. Thinking this through, we might have a race if MeetEcho calls `api_add_session_attendees` with the user pk, after the user GETs the form but before they POST. Doing the `can_add` check before accepting the POST means that it will now fail the check (there's now an existing `Attendance` record).

But I also don't understand `defaults` in this context. Are you suggesting something like what appears in `nomcom.views.volunteer`?

jennifer-richards replied:
> we might have a race if MeetEcho calls api_add_session_attendees with the user pk, after the user GETs the form but before they POST. Doing the can_add check before accepting the POST means that it will now fail the check (there's now an existing Attendance record).

That is the race I'm thinking about. Adding the `can_add` check to the POST significantly narrows the window, but there's still a race because MeetEcho's API call might happen just after testing `can_add`. What I'm suggesting is like the volunteer view you mention, specifically:
```python
session.attended_set.get_or_create(person=person, defaults={"origin": "self declared"})
```
The defaults are used only when creating a record because wasn't found - this is equivalent to trying a `.filter(person=person).get()`, then if it fails, creating a person with `(person=person, origin="self declared")`. It has the advantage that it enforces that check in the DB, so it closes the race completely.

I'm assuming that the other side of this also has a guard against this race. It might not, in which case in this scheme I think it should get a guard.

Alternatively, we could just allow duplicates. That might be easier. We'd need to be careful to de-dupe when counting and to be careful about priority when displaying the origin - I'd assume that having both a "datatracker" and a "self declared" would be treated as "datatracker" for most purposes since that is the more reliable source.

rjsparks replied:
I would prefer to avoid duplicates if at all possible.

pselkirk replied:
Apparently that works. Done and tested.

pselkirk replied:
The `Attended` model has `unique_together = (('person', 'session'),)`, so presumably django would raise an exception if we try to add a duplicate.

jennifer-richards replied:
Ah, that's great.

It'll turn into a 500 response unless the `IntegrityError` is handled , though. IMO better to race-proof the code. The only reason I can think of to allow this to fail on that is if we wanted to catch it and display a "You were already marked 'attended'" error, but that doesn't seem necessary.

rjsparks replied:
However, I don't think the code's using the get_or_create(...,defaults=) pattern yet, and it would be much more robust if it did so. What's the reasning for not doing that?

pselkirk replied:
> However, I don't think the code's using the get_or_create(...,defaults=) pattern yet, and it would be much more robust if it did so. What's the reasoning for not doing that?

I did that in 1e7995c67c, sorry if it wasn't clear.

jennifer-richards replied:
Thanks! I didn't review the updates closely enough (or notice the big "Outdated" badge on the comment...)

pselkirk committed:
fix: Address review comments

pselkirk committed:
fix: Don't try to generate empty bluesheets

pselkirk committed:
refactor: Complete rewrite of bluesheet.html

pselkirk committed:
Merge branch 'main' into fix-6898-6454

jennifer-richards reviewed (commented):
One remaining question: what happens if `save_bluesheet()` returns an error?

I came across it noticing that `generate_bluesheet()` now returns `None` on one branch and returns the outcome of `save_bluesheet()` on the other.

pselkirk committed:
refactor: Fill in a few gaps, close a few holes

- Rename the live "bluesheet" to "attendance", add some explanatory text.
- Add attendance links in materials view and pre-finalized proceedings view.
- Don't allow users to add themselves after the corrections cutoff date.

pselkirk committed:
fix: Report file-save errors to caller

jennifer-richards reviewed (dismissed):
My concerns are addressed.

rjsparks commented on ietf/meeting/views.py:
```diff
@@ -2516,6 +2527,59 @@ def add_session_drafts(request, session_id, num):
+    meeting = session.meeting
+    return [{'name':attended.person.name, 'affiliation':affiliation(meeting, attended.person)} for attended in attendance]
+
+def session_attendance(request, session_id, num):
```
This needs to not allow someone to say "I attended" before they attended. It should probably not let people in at all until at least one Attended record exists for a session. That's counting on a side-effect since the only other way such a record would be created is MeetEcho's upload right now, but we should take advantage of that for now - otherwise we have to define when the view should open based on some calculation of "the session is over".

jennifer-richards replied:
Using that "side effect" has the nice property that it mostly prevents impatient people from self-attesting their attendance before MeetEcho uploads their records. I.e., I think it's pretty close to the right signal to open this mechanism up.

rjsparks reviewed (changes requested):
The new attendance view needs slightly tighter gating.

rjsparks commented on ietf/meeting/views.py:
```diff
@@ -2516,6 +2527,59 @@ def add_session_drafts(request, session_id, num):
+
+    attendance = Attended.objects.filter(session=session)
+    meeting = session.meeting
+    return [{'name':attended.person.name, 'affiliation':affiliation(meeting, attended.person)} for attended in attendance]
```
This should use person.plain_name()

rjsparks commented on ietf/templates/meeting/attendance.html:
```diff
@@ -0,0 +1,41 @@
+    <h1>
+        Attendance for {{session}}
+    </h1>
+    <div class="alert alert-info my-3">
```
We should put something like this here:
"If the affiliation listed here needs to be updated, request the change using support@ietf.org. Note which sessions you are wanting to change in your request."

alexamirante commented on ietf/meeting/views.py:
```diff
@@ -2516,6 +2527,59 @@ def add_session_drafts(request, session_id, num):
+    data = bluesheet_data(session)
+    if not data:
+        return
+    text = render_to_string('meeting/bluesheet.txt', {
```
@rjsparks yes, we can send join times as well if needed.

pselkirk committed:
fix: Address review comments

pselkirk committed:
fix: typo

rjsparks commented on ietf/meeting/views.py:
```diff
@@ -2516,6 +2527,65 @@ def add_session_drafts(request, session_id, num):
+
+    cor_cut_off_date = session.meeting.get_submission_correction_date()
+    today_utc = date_today(datetime.timezone.utc)
+    try:
+        person = request.user.person
+        was_there = Attended.objects.filter(session=session, person=person).exists()
+        can_add = today_utc <= cor_cut_off_date and MeetingRegistration.objects.filter(meeting=session.meeting, person=person).exists() and Attended.objects.filter(session=session).exists() and not was_there
+    except AttributeError:
+        was_there = can_add = False
```
I think something like
`if user.is_authenticated and user.person is not None:`
would be a lot better

rjsparks commented:
This became feat/sess-apis
